### PR TITLE
tests: CLOSE_WAIT test should wait for pods to be ready

### PR DESCRIPTION
**Pull request by justinsb**
We sometimes observe test flakes in the CLOSE_WAIT test.  The problem
seems to be that there is a time-sensitive portion of the test where
we try to observe a socket in a particular state.  However, the pod
from which we try to observe this state may not have launched yet.

Wait for all pods to be ready before entering the time-critical
portion of the test.

/kind failing-test
/kind flake

```release-note
NONE
```

### CODE REVIEW

**Comment by justinsb** [comment]
Example of this flake: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/e2e-kops-grid-calico-u2004-k19-docker/1361197062846156800

Observe the long (2 minute) delay in the first invocation of conntrack:

```
ï¿½[1mSTEPï¿½[0m: Launching a server daemon on node ip-172-20-58-31.eu-central-1.compute.internal (node ip: 172.20.58.31, image: k8s.gcr.io/e2e-test-images/agnhost:2.20)
ï¿½[1mSTEPï¿½[0m: Launching a client connection on node ip-172-20-37-245.eu-central-1.compute.internal (node ip: 172.20.37.245, image: k8s.gcr.io/e2e-test-images/agnhost:2.20)
ï¿½[1mSTEPï¿½[0m: Checking conntrack entries for the timeout
Feb 15 06:35:59.376: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://api.e2e-kops-calico-u2004-k19-docker.test-cncf-aws.k8s.io --kubeconfig=/tmp/kops090787714/kubeconfig --namespace=network-5667 exec e2e-net-exec -- /bin/sh -x -c conntrack -L -f ipv4 -d 172.20.58.31| grep -m 1 'CLOSE_WAIT.*dport=11302' '
Feb 15 06:38:12.257: INFO: rc: 1
Feb 15 06:38:12.342: INFO: failed to obtain conntrack entry:  error running /workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://api.e2e-kops-calico-u2004-k19-docker.test-cncf-aws.k8s.io --kubeconfig=/tmp/kops090787714/kubeconfig --namespace=network-5667 exec e2e-net-exec -- /bin/sh -x -c conntrack -L -f ipv4 -d 172.20.58.31| grep -m 1 'CLOSE_WAIT.*dport=11302' :
Command stdout:

stderr:
+ grep -m 1 CLOSE_WAIT.*dport=11302
+ conntrack -L -f ipv4 -d 172.20.58.31
conntrack v1.4.5 (conntrack-tools): 3 flow entries have been shown.
command terminated with exit code 1

error:
exit status 1
Feb 15 06:38:12.377: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://api.e2e-kops-calico-u2004-k19-docker.test-cncf-aws.k8s.io --kubeconfig=/tmp/kops090787714/kubeconfig --namespace=network-5667 exec e2e-net-exec -- /bin/sh -x -c conntrack -L -f ipv4 -d 172.20.58.31| grep -m 1 'CLOSE_WAIT.*dport=11302' '
Feb 15 06:38:33.568: INFO: rc: 1
Feb 15 06:38:33.596: INFO: failed to obtain conntrack entry:  error running /workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://api.e2e-kops-calico-u2004-k19-docker.test-cncf-aws.k8s.io --kubeconfig=/tmp/kops090787714/kubeconfig --namespace=network-5667 exec e2e-net-exec -- /bin/sh -x -c conntrack -L -f ipv4 -d 172.20.58.31| grep -m 1 'CLOSE_WAIT.*dport=11302' :
Command stdout:

stderr:
+ conntrack -L -f ipv4 -d 172.20.58.31
+ grep -m 1 CLOSE_WAIT.*dport=11302
conntrack v1.4.5 (conntrack-tools): 2 flow entries have been shown.
command terminated with exit code 1

error:
exit status 1
Feb 15 06:38:33.629: FAIL: no conntrack entry for port 11302 on node 172.20.58.31
```

Later we can see that the pod was not ready until 38:33:

```
ï¿½[1mSTEPï¿½[0m: Collecting events from namespace "network-5667".
ï¿½[1mSTEPï¿½[0m: Found 12 events.
Feb 15 06:38:33.786: INFO: At 2021-02-15 06:35:39 +0000 UTC - event for e2e-net-exec: {kubelet ip-172-20-37-245.eu-central-1.compute.internal} Pulling: Pulling image "k8s.gcr.io/build-image/debian-iptables:v12.1.2"
Feb 15 06:38:33.786: INFO: At 2021-02-15 06:35:44 +0000 UTC - event for e2e-net-exec: {kubelet ip-172-20-37-245.eu-central-1.compute.internal} Created: Created container e2e-net-exec
Feb 15 06:38:33.786: INFO: At 2021-02-15 06:35:44 +0000 UTC - event for e2e-net-exec: {kubelet ip-172-20-37-245.eu-central-1.compute.internal} Pulled: Successfully pulled image "k8s.gcr.io/build-image/debian-iptables:v12.1.2" in 5.093070619s
Feb 15 06:38:33.786: INFO: At 2021-02-15 06:35:45 +0000 UTC - event for e2e-net-exec: {kubelet ip-172-20-37-245.eu-central-1.compute.internal} Started: Started container e2e-net-exec
```

**Comment by justinsb** [comment]
/retest

That's ... a lot of failures that should be unrelated to this patch :-(

**Comment by BenTheElder** [comment]
/cc @aojea 
one of our perennial "favorite" flakes

File: test/e2e/network/kube_proxy.go [hunk 1]
```diff
@@ -182,6 +182,11 @@ var _ = SIGDescribe("KubeProxy", func() {
 			kubeProxyE2eImage))
 		fr.PodClient().CreateSync(serverPodSpec)
 
+		// The host exec pod should be ready before we start the test, as it is time sensitive
+		// (if the host exec pod takes a long time to be ready, that could cause spurious timeouts)
+		if readyErr := e2epod.WaitForPodsReady(fr.ClientSet, fr.Namespace.Name, hostExecPod.Name, 0); readyErr != nil {
```
Comment (aojea) [inline test/e2e/network/kube_proxy.go:187]: :thinking: I thought that the `CreateSync()` with the `WaitForPodNameRunningInNamespace` was enough
https://github.com/kubernetes/kubernetes/blob/5a382b2d836fb7f57fd1631251651d00c802645b/test/e2e/framework/pods.go#L99-L108

Comment (aojea) [reply test/e2e/network/kube_proxy.go:187]: if that is not the case then I think we should modify the CreateSync os everybody using it got the right behaviour

Comment (aojea) [reply test/e2e/network/kube_proxy.go:187]: what do you thing about replacing the `e2epod.WaitForPodNameRunningInNamespace`  in `CreateSync` by 
```
// WaitTimeoutForPodReadyInNamespace waits the given timeout diration for the
// specified pod to be ready and running.
func WaitTimeoutForPodReadyInNamespace(c clientset.Interface, podName, namespace string, timeout time.Duration) error {
	return wait.PollImmediate(poll, timeout, podRunningAndReady(c, podName, namespace))
}
```
or better do this in another PR?

Comment (aojea) [reply test/e2e/network/kube_proxy.go:187]: scratch this, it already has a WaitForReady in another places ...

Comment (aojea) [reply test/e2e/network/kube_proxy.go:187]: https://github.com/kubernetes/kubernetes/pull/99147

**Comment by aojea** [comment]
@justinsb since this merged before yours https://github.com/kubernetes/kubernetes/pull/99147 it should be ok to close this, right?
Anyway, thanks for the detailed analysis and pinpointing the problem
https://github.com/kubernetes/kubernetes/pull/99097#issuecomment-779293579

**Comment by aojea** [comment]
> @justinsb since this merged before yours #99147 it should be ok to close this, right?
> Anyway, thanks for the detailed analysis and pinpointing the problem
> [#99097 (comment)](https://github.com/kubernetes/kubernetes/pull/99097#issuecomment-779293579)
per ^^^
/close

**Pull request closed without merging**